src/shmem_internal.h

```diff
@@ -247,46 +247,66 @@ extern hwloc_topology_t shmem_internal_topology;
-                                                   shmem_internal_heap_length);         \
-        if (! (ptr_base >= shmem_internal_heap_base && ptr_base < heap_ext)) {          \
-            RAISE_ERROR_MSG("Argument \"%s\" is not in symm. heap (%p), [%p..%p)\n",    \
-                    #ptr_in, ptr_base, shmem_internal_heap_base, heap_ext);             \
-        }                                                                               \
+#define SHMEM_ERR_CHECK_SYMMETRIC(ptr_in, len)                                                \
+    do {                                                                                      \
+        const void *ptr_base = (void*)(ptr_in);                                               \
+        const void *ptr_ext  = (void*)((uint8_t *) (ptr_base) + (len));                       \
+        const void *data_ext = (void*)((uint8_t *) shmem_internal_data_base +                 \
+                                                   shmem_internal_data_length);               \
+        const void *heap_ext = (void*)((uint8_t *) shmem_internal_heap_base +                 \
+                                                   shmem_internal_heap_length);               \
+        if (len == 0) {                                                                       \
+            break; /* Skip this check when the length is 0 */                                 \
+        }                                                                                     \
+        else if (ptr_base >= shmem_internal_data_base && ptr_base < data_ext) {               \
+            if (ptr_ext > data_ext) {                                                         \
+                RAISE_ERROR_MSG("Argument \"%s\" [%p..%p) exceeds "                           \
+                                "sym. data region [%p..%p)\n", #ptr_in, ptr_base,             \
+                                ptr_ext, shmem_internal_data_base, data_ext);                 \
+            }                                                                                 \
+        }                                                                                     \
+        else if (ptr_base >= shmem_internal_heap_base && ptr_base < heap_ext) {               \
+            if (ptr_ext > heap_ext) {                                                         \
+                RAISE_ERROR_MSG("Argument \"%s\" [%p..%p) exceeds "                           \
+                                "sym. heap region [%p..%p)\n", #ptr_in, ptr_base,             \
+                                ptr_ext, shmem_internal_heap_base, heap_ext);                 \
+            }                                                                                 \
+        }                                                                                     \
+        else if (shmem_external_heap_base) {                                                  \
```

Comment:
what kind of devices could provide an external heap? GPU's would be my first thought, but I was under the impression that SOS doesn't support using their memory.

Reply:
In the case of Intel® SHMEM, we are using the external heap to allow SOS to interact with data in device (GPU) memory. SOS supports this usage when configured with **--enable-ofi-hmem**.